src/api.rs

```diff
@@ -256,17 +272,95 @@ impl Api {
+            hashes = hashes_query,
+            tag = percent_encode(tag.as_bytes(), NON_ALPHANUMERIC)
+        );
+
+        let response = self
             .client
-            .get(&addr)
-            .headers(self.make_headers()?)
+            .post(&addr)
+            .header(header::CONTENT_TYPE, "application/x-www-form-urlencoded")
+            .body(body)
             .send()
-            .await?
-            .bytes()
             .await?;
-        dbg! {res};
+
+        if response.status() != StatusCode::OK {
+            bail!(
+                "Cannot tag torrents: {status} {body}",
+                status = response.status(),
+                body = response.text().await?,
+            );
+        }
+
+        Ok(())
+    }
+
+    pub async fn tag_error_trackers(&self, tag: &str, all: bool) -> Result<(), anyhow::Error> {
+        let torrents = self.get_torrent_list().await.unwrap();
+
+        let mut handles = vec![];
+        let sem = Arc::new(tokio::sync::Semaphore::new(20));
+        for torrent in torrents {
+            let api = self.clone();
+            let permit = Arc::clone(&sem).acquire_owned().await;
+            handles.push(tokio::spawn(async move {
+                let _permit = permit;
+                let hash = torrent.hash();
+                let name = torrent.name();
+                let trackers = api.get_trackers(hash).await.expect(
+                    &format!(
+                        "Cannot get trackers from: [{hash}] {name}",
+                        hash = hash,
+                        name = name
+                    )[..],
+                );
+
+                for tracker in trackers {
+                    match tracker.status() {
+                        TrackerStatus::TrackerDisabled | TrackerStatus::Working => continue,
+                        _ => {
+                            if tracker.msg().len() > 0 || all {
+                                println!(
+                                    "[{hash}]({status:?}) {name} ({msg})",
+                                    hash = hash,
+                                    status = tracker.status(),
+                                    name = name,
+                                    msg = tracker.msg()
+                                );
+                                return Some(hash.clone());
+                            } else {
+                                continue;
+                            }
+                        }
+                    }
+                }
+
+                None
+            }));
+        }
+        let results = futures::future::join_all(handles).await;
+
+        let results: Vec<String> = results
+            .into_iter()
+            .filter_map(Result::ok)
+            .filter(|x| x.is_some())
+            .map(|x| x.unwrap().hash)
+            .collect();
+
+        let hashes: Vec<&str> = results.iter().map(|s| &**s).collect();
+
+        self.add_tag(hashes, tag).await?;
 
         Ok(())
     }
```

Comment:
I'm on the fence about adding this here because its not actually wrapping any api endpoints its more application logic. Is there any reason this can't be moved into your application instead of the crate? For example, if you are `println!`-ing things in here anyone using the library will have their stdout polluted by info they don't need.

Additionally, you use `tokio` as a dependency in this method whereas everywhere else it could be a dev dependency that is only required for running tests. This would lock anyone using this library into using tokio or being forced to run multiple executors. A younger and wiser me would have simply made it a dev dependency to avoid this issue so its partly my fault.